File: Core/src/WebView/SAWebViewFaviconLoader.m
/*
 * This file is part of Adblock Plus <https://adblockplus.org/>,
 * Copyright (C) 2006-present eyeo GmbH
 *
 * Adblock Plus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * Adblock Plus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "SAWebViewFaviconLoader.h"
#import "SAContentWebView.h"
#import "Settings.h"

#import <KittCore/KittCore-Swift.h>

static NSString *const FAVICON_DEFAULT_PATH = @"/favicon.ico";

static NSString *const ICON_LINK_REL_APPLE = @"apple-touch-icon";

static NSArray *_scenarioSelectors;
static dispatch_queue_t _loadingQueue;

@interface FaviconObject : NSObject <FaviconFacade>

@property (nonatomic, strong) NSString *iconUrl;
@property (nonatomic, strong) NSData *iconData;
@property (nonatomic, strong) NSNumber *size;

@end

@implementation FaviconObject

@end

@interface SAWebViewFaviconLoader () <NSURLConnectionDataDelegate> {
    NSUInteger _selectorNextIndex;
    FaviconGroup *_faviconGroup;
    NSURLRequest *_currentRequest;
    NSURL *_requestedFaviconURL;
    NSInteger _requestedFaviconSize;
}

@property (nonatomic, weak) id<FaviconLoadingDelegate> delegate;
@property (nonatomic, strong) BrowserHistoryManager *historyManager;
@end

@implementation SAWebViewFaviconLoader

+ (void)initialize
{
    _scenarioSelectors = @[ NSStringFromSelector(@selector(URLFaviconStringFromDOMOfServerURL:)),
        NSStringFromSelector(@selector(URLFaviconStringWithDefaultServerURL:)),
        NSStringFromSelector(@selector(URLFaviconStringWithModifiedServer:)) ];
    _loadingQueue = dispatch_queue_create("com.kitt.faviconloading", DISPATCH_QUEUE_SERIAL);
}

+ (dispatch_queue_t)loadingQueue
{
    return _loadingQueue;
}

- (BrowserHistoryManager *)historyManager
{
    return _historyManager;
}

- (instancetype)initWithDelegate:(id<FaviconLoadingDelegate>)delegate
                  historyManager:(BrowserHistoryManager *)historyManager
{
    if (self = [super init]) {
        _delegate = delegate;
        _historyManager = historyManager;
    }
    return self;
}

- (void)startFaviconLoadingFromURL:(NSURL *)currentFaviconURL
                          withSize:(NSUInteger)size
                   andFaviconGroup:(FaviconGroup *)faviconGroup
{
    if (faviconGroup.resolved) {
        return;
    }
    dispatch_suspend(_loadingQueue);
    _faviconGroup = faviconGroup;
    _currentRequest = faviconGroup.request;
    _requestedFaviconSize = size;
    _selectorNextIndex = 0;
    // doesn't need dispatch to main thread. Connection start is threadsafe and result callback
    // is queued on main thread anyway
    [self startConnectionWithURL:currentFaviconURL];
}

- (void)startFaviconLoadingWith:(NSURLRequest *)currentRequest
{
    _currentRequest = currentRequest;
    _requestedFaviconSize = 0;
    _selectorNextIndex = 0;
    [self connectionCompletedWithFaviconData:nil];
}

- (void)connectionCompletedWithFaviconData:(NSData *)data
{
    if (data) {
        if (_requestedFaviconSize == 0) {
            // no specific favicon size was requested, find out
            UIImage *iconImage = [UIImage imageWithData:data];
            if (!iconImage) {
                [_delegate setCurrentFavicon:nil];
                dispatch_resume(_loadingQueue);
                return;
            }
            _requestedFaviconSize = iconImage.size.width;
        }

        NSMutableArray<NSURL *> *urls = [NSMutableArray array];
        if (_currentRequest.URL) {
            [urls addObject:_currentRequest.URL];
        }
        if (_currentRequest.originalURL) {
            [urls addObject:_currentRequest.originalURL];
        }
        if (_historyManager && [urls count] > 0) {
            UrlIcon *icon = [_historyManager attach:data
                                        fromIconURL:_requestedFaviconURL
                                           withSize:_requestedFaviconSize
                                             toURLs:urls];
            _faviconGroup.favicon = icon;
            [_delegate setCurrentFavicon:icon];
        } else {
            FaviconObject *object = [[FaviconObject alloc] init];
            object.iconData = data;
            object.iconUrl = _requestedFaviconURL.absoluteString;
            object.size = [[NSNumber alloc] initWithInteger:_requestedFaviconSize];
            _faviconGroup.favicon = object;
            [_delegate setCurrentFavicon:object];
        }

        dispatch_resume(_loadingQueue);
    } else if (_selectorNextIndex == [_scenarioSelectors count]) {
        // all strategies failed
        [_delegate setCurrentFavicon:nil];
        dispatch_resume(_loadingQueue);
    } else {
        // continue with next
        SEL selector = NSSelectorFromString(_scenarioSelectors[_selectorNextIndex++]);
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
        NSString *scenarioURLString = [self performSelector:selector withObject:_currentRequest.URL];
#pragma clang diagnostic pop
        if (scenarioURLString) {
            NSURL *scenarioURL = [NSURL URLWithString:scenarioURLString];
            [self startConnectionWithURL:scenarioURL];
        } else {
            [self connectionCompletedWithFaviconData:nil];
        }
    }
}

#pragma mark - Static

+ (NSString *)URLStringFromValidatedFaviconURL:(NSURL *)url withCurrentURL:(NSURL *)currentURL
{
    NSString *ret = url.absoluteString;
    if ([url.host length] == 0) {
        // URL without host, resolve against host and make absolute again
        url = [NSURL URLWithString:ret relativeToURL:currentURL];
        ret = url.absoluteString;
    } else if ([ret hasPrefix:@"//"]) {
        // Absolute URL (incl. hostname).
        // Icons can apparently be declared without 'protocol:'
        // http://stackoverflow.com/a/16844961
        ret = [NSString stringWithFormat:@"%@:%@",
                        url.scheme, ret];
    }
    return ret;
}

#pragma mark - Private

- (void)startConnectionWithURL:(NSURL *)url
{
    _requestedFaviconURL = url;
    NSMutableURLRequest *iconRequest = [NSMutableURLRequest requestWithURL:url];
    // all outgoing requests better have a valid UA
    [iconRequest setValue:[Settings defaultWebViewUserAgent] forHTTPHeaderField:@"User-Agent"];
    [NSURLConnection sendAsynchronousRequest:iconRequest
                                       queue:[NSOperationQueue mainQueue]
                           completionHandler:^(NSURLResponse *response, NSData *data, NSError *connectionError) {
                               // handler callbacks should be on main thread because we asked for it, but better check
                               NSAssert([NSThread isMainThread], @"NSURLConnection.sendAsynchronousRequest completion not queued on main thread");
                               if (connectionError) {
                                   [self connectionCompletedWithFaviconData:nil];
                                   return;
                               }
                               if ((response.expectedContentLength > 0L) && [response isKindOfClass:[NSHTTPURLResponse class]]) {
                                   NSHTTPURLResponse *iconResponse = (NSHTTPURLResponse *)response;
                                   NSString *contentType = [iconResponse allHeaderFields][@"Content-Type"];
                                   if ((iconResponse.statusCode == 200) &&
                                       // There are sites returning 200 with text/html. Ignore those
                                       ([contentType hasPrefix:@"image/"] ||
                                           // Slashdot be damned. Serving ico with text/plain mime is lunatic.
                                           [contentType hasPrefix:@"text/plain"])) {
                                       [self connectionCompletedWithFaviconData:data];
                                       return;
                                   }
                               }
                               [self connectionCompletedWithFaviconData:nil];
                           }];
}

@end

File: Core/src/WebView/SAWebViewFaviconLoader.swift
/*
 * This file is part of Adblock Plus <https://adblockplus.org/>,
 * Copyright (C) 2006-present eyeo GmbH
 *
 * Adblock Plus is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * Adblock Plus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
 */

import Foundation

extension SAWebViewFaviconLoader {
    @objc
    public func verifyFaviconWith(_ currentRequest: URLRequest) {
        assert(Thread.isMainThread)

        if delegate == nil {
            return
        }

        guard let currentURL = currentRequest.url else {
            return
        }

        let currentURLs: [URL]
        if let url = currentRequest.originalURL {
            currentURLs = [currentURL, url]
        } else {
            currentURLs = [currentURL]
        }

        let icon = historyManager()?.faviconFor(urls: currentURLs)

        if icon == nil {
            // no icon yet, look for some
            let loadingQueue = type(of: self).loadingQueue()
            loadingQueue.async {[weak self] in
                loadingQueue.suspend()
                DispatchQueue.main.async {
                    // verify again that there is still no icon
                    if self?.historyManager()?.faviconFor(urls: currentURLs) != nil {
                        // icon has landed meanwhile
                        loadingQueue.resume()
                    } else if let start = self?.startFaviconLoading(with:) {
                        start(currentRequest)
                    } else {
                        loadingQueue.resume()
                    }
                }
            }
        }
    }

    /// [delegate faviconData] will be called when finished
    func load(_ favicons: [FaviconSource], fromRequest currentRequest: URLRequest) {
        guard let currentURL = currentRequest.url else {
            return
        }

        let currentURLs: [URL]
        if let url = currentRequest.originalURL {
            currentURLs = [currentURL, url]
        } else {
            currentURLs = [currentURL]
        }

        let icon = historyManager()?.faviconFor(urls: currentURLs)

        var favicons = favicons

        if let icon = icon, icon.size != 0 {
            favicons = favicons.filter {
                guard let otherSize = icon.size?.int16Value else {
                    return false
                }
                return $0.size > otherSize || $0.type == faviconAppleTouch
            }
        }

        guard favicons.count != 0 else {
            // dispatch to obey the declared asynchronicity
            let delegate = self.delegate
            DispatchQueue.main.async {
                delegate?.setCurrentFavicon(icon)
            }
            return
        }

        favicons.sort { icon1, icon2 in
            let value1 = icon1.type == faviconAppleTouch ? 1 : 0
            let value2 = icon2.type == faviconAppleTouch ? 1 : 0

            if value1 != value2 {
                return value1 > value2
            } else {
                return icon1.size > icon2.size
            }
        }

        let faviconGroup = FaviconGroup(request: currentRequest)

        for favicon in favicons {
            let validatedURLString = type(of: self).urlString(fromValidatedFaviconURL: favicon.url as URL, withCurrentURL: currentURL)
            if let validatedURLString = validatedURLString, let currentFaviconURL = URL(string: validatedURLString) {
                let loadingQueue = type(of: self).loadingQueue()
                loadingQueue.async {[weak self] in
                    self?.startFaviconLoading(from: currentFaviconURL, withSize: UInt(favicon.size), andFaviconGroup: faviconGroup)
                }
            }
        }
    }

    // MARK: Favicon gathering

    // Scenario 1
    @objc
    public func URLFaviconStringFromDOMOfServerURL(_ serverURL: URL) -> String? {
        // Apple specific first, fall back to unspecific
        let urlString: String
        if let url = queryDOMWithAppleSpecific(true) {
            urlString = url
        } else if let url = queryDOMWithAppleSpecific(false) {
            urlString = url
        } else {
            // DOM doesn't specify any icon, bail out
            return nil
        }

        if let url = URL(string: urlString), let currentURL = currentRequest?.url {
            return type(of: self).urlString(fromValidatedFaviconURL: url, withCurrentURL: currentURL)
        } else {
            return nil
        }
    }

    // Scenario 2
    @objc
    public func URLFaviconStringWithDefaultServerURL(_ serverURL: URL) -> String? {
        // NSURL has no way to replace just the path
        // NSURLComponents (aka fully mutable NSURL) is iOS7+
        // So we must reconstruct as string
        if let scheme = serverURL.scheme, let host = serverURL.host {
            return "\(scheme)://\(host)\(optionalFormattedPort(of: serverURL))\(faviconDefaultPath)"
        } else {
            return nil
        }
    }

    // Scenario 3
    @objc
    public func URLFaviconStringWithModifiedServer(_ serverURL: URL) -> String? {
        guard let domainLevels = serverURL.host?.components(separatedBy: "."), domainLevels.count > 1 else {
            // not a regular URL, not a subject for this scenario
            return nil
        }
        guard let scheme = serverURL.scheme else {
            return nil
        }
        return "\(scheme)://www.\(domainLevels[domainLevels.count - 2])"
            + ".\(domainLevels[domainLevels.count - 1])\(optionalFormattedPort(of: serverURL))\(faviconDefaultPath)"
    }

    func queryDOMWithAppleSpecific(_ appleSpecific: Bool) -> String? {
        if let delegate = delegate {
            let selector = appleSpecific ?
                "link[rel='\(faviconAppleTouch)']" :
            "link[rel='shortcut icon'], link[rel='icon']"
            // wrap in IIFE to prevent interaction with global scope
            let query = "(function(){var s=document.querySelector(\"\(selector)\");return s?s.getAttribute('href'):null})();"
            let result = delegate.string(fromEvalJS: query)
            // return nil if JS returned empty string (which equals to "not found")
            return result.isEmpty ? nil : result
        } else {
            return nil
        }
    }

    // NSURL doesn't have a method to obtain complete hostname with port. So if
    // just a partial URL change is needed (like path only), port needs to be
    // reinserted. This function does it in a way that the result can be used
    // unconditionally
    // @return port number prefixed with separator
    // @return empty string if serverURL has no port
    func optionalFormattedPort(of serverURL: URL) -> String
    {
        if let port = serverURL.port {
            return ":\(port)"
        } else {
            return ""
        }
    }
}

private let faviconDefaultPath = "/favicon.ico"

private let faviconAppleTouch = "apple-touch-icon"

open class FaviconGroup: NSObject {
    @objc open let request: URLRequest
    @objc open var favicon: FaviconFacade?

    init(request: URLRequest) {
        self.request = request
    }

    @objc open var resolved: Bool {
        return favicon != nil
    }
}

struct FaviconSource {
    let url: URL
    let type: String?
    let size: Int16

    init?(object: Any) {
        guard let properties = object as? [String: Any] else {
            Log.warn("Unsupported type of properties")
            return nil
        }

        assert(properties.count == 3, "favicon event parameters expected to be array of 3 elements")

        guard let urlString = properties["href"] as? String, let url = URL(string: urlString) else {
            Log.warn("Favicon without URL")
            return nil
        }

        var biggestSize = Int16(0)

        if let sizes = properties["sizes"] as? String, !sizes.isEmpty {
            // http://www.w3.org/html/wg/drafts/html/master/semantics.html#attr-link-sizes
            // The format is "any" or a space-separated list of nonegative integer tuples
            // e.g. "45x45 72x72 120x120"
            for size in sizes.components(separatedBy: CharacterSet.whitespaces) {
                let keyword = size.lowercased()

                if keyword == "any" {
                    biggestSize = Int16.max
                    break
                }

                let dimensions = keyword.components(separatedBy: "x")
                if dimensions.count != 2 {
                    // does not contain 2 tokens separated by "x"
                    continue
                }

                let strDimX = dimensions[0]
                let strDimY = dimensions[1]

                if strDimX.hasPrefix("0") || strDimY.hasPrefix("0") {
                    // the numbers must not start with zero
                    continue
                }

                if let dimX = Int16(strDimX), let dimY = Int16(strDimY) {
                    biggestSize = max(biggestSize, max(dimX, dimY))
                }
            }
        }

        self.url = url
        self.type = properties["rel"] as? String
        self.size = biggestSize
    }
}
